### DISCUSSION TITLE
Using DevContainers and VSCode for python and django

### DISCUSSION THREAD

**Discussion by thclark** [octue/.github#16] [Python Libraries and Applications]
Use this discussion to record tips and tricks for how to set up devcontainers and/or VSCode

**Comment by thclark** [comment]
## Debugging / Launch profiles in VSCode

In VSCode, you want to be able to use pytest to debug your tests, and (if developing a django application or library) your django application itself.

This will involve setting up your test framework with pytest, but you should also set up launch profiles in VSCode. To do that:

In [django-gcp](https://github.com/octue/django-gcp) we have `pytest` set up and it's easy enough to launch it from the command line.

But, we want to use VSCode's built in debugging. To do that we create launch profiles:

**In `.vscode/launch.json`:**
```js
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python: Debug Tests",
            "type": "debugpy",
            "request": "launch",
            "program": "${file}",
            "django": true,
            "purpose": ["debug-test"],
            "console": "integratedTerminal",
            "justMyCode": true, // Make this false to set breakpoints in third party code
            "env": {"PYTEST_ADDOPTS": "--no-cov"}
        },
        {
            "name": "Python: Django",
            "type": "debugpy", 
            "request": "launch",
            "program": "${workspaceFolder}/manage.py",
            "args": [
                "runserver"
            ],
            "django": true,
            "justMyCode": true // Make this false to set breakpoints in third party code
        }
    ]
}
```

Launching django from the debug tools in the sidebar now launches django in debug mode. You can make any requests to your local server you like and hte execution will stop at any breakpoints in your code.

![Screenshot 2024-10-31 at 14 01 21](https://github.com/user-attachments/assets/5556ab82-dab3-40f3-a25c-c3d945d725d3)

**Comment by thclark** [comment]
## Migrating from ruff-lsp

Ruff settings have been updated, so devcontainers using ruff need to switch from:
```js
{
  "customizations": {
    "vscode": {
      "settings": {
        // ...
        "ruff.format.args": ["--line-length", "120"],
```
to
```js
{
  "customizations": {
    "vscode": {
      "settings": {
        // ...
        "ruff.lineLength":120,
```

[More information available here](https://docs.astral.sh/ruff/editors/migration/#examples)

**Comment by cortadocodes** [comment]
# Update for using non-VSCode devcontainer images
This line is in many of our `devcontainer.json` files:
```
// Comment out to connect as root instead. More info: https://aka.ms/vscode-remote/containers/non-root.
"remoteUser": "vscode",
```

## Problem
It (usually?) works if you're using a microsoft based image, but if you need to use something else, the devcontainer won't launch, citing `no users found`. This is at least partly because there is no `vscode` user.

## Solution
The solution is to set the `remoteUser` to `root`. However, I've just found that the comment above is misleading - you have to set `remoteUser` to `root` when you're not using the `vscode` user, not just comment out the line:
```
// Set to "root" if using a non-microsoft base image. More info: https://aka.ms/vscode-remote/containers/non-root.
"remoteUser": "root",
```
